### DISCUSSION TITLE
Upcoming Breaking Changes

### DISCUSSION THREAD

**Discussion by pml-lang** [pml-lang/pml-companion#97] [Ideas]
# Upcoming Breaking Changes

The following breaking changes are planned to be implemented in the upcoming PML major version 4.0.0.

Feedback before implementing them is welcome.

## Attributes Containing Text Replaced With Child-Nodes

A few nodes use attributes to define _text rendered in the document_.

Instead of using attributes, it is better to use child-nodes, for the following reasons:

- Using an attribute to define text precludes the text from being formatted/styled, because attributes can only contain plain text (inline nodes are not supported).
- Attributes should only be used to contain _meta data_ (such as as the width and height of an image). They should not be used to define text rendered in the document.

The following nodes are affected: `link`, `xref`, `quote`, `admon`.

### Node `link`

Attribute `text` will be removed. The link text must be contained in the node.

Hence, this code:

    [link url=http://example.com text="an example"]

must be replaced by:

    [link (url=http://example.com) an example]

The text can then be formatted/styled by using inline nodes:

    [link (url=http://example.com) a [i striking] example]

### Node `xref`

Attribute `text` will be removed. The text must be contained in the node.

Moreover attribute `node_id` will be renamed to `ref_id`.

Hence, this code:

    [xref node_id=intro text="chapter Introduction"]

must be replaced by:

    [xref (ref_id=intro) chapter Introduction]

### Node `quote`

Attribute `source` will be removed. A new node named `q_source` is used within node `quote` to define the source of the quote.

Hence, this code:

    [quote (source="Scott Adams")
        Simplicity transforms ordinary into amazing.
    ]

must be replaced by:

    [quote
        Simplicity transforms ordinary into amazing.
        [q_source Scott Adams]
    ]

### Node `admon`

Attribute `label` will be removed. A new node named `a_label` is used within node `admon` to define the label of the admonition.

Hence, this code:

    [admon (label=Tip)
        You can also ...
    ]

must be replaced by:

    [admon [a_label Tip]
        You can also ...
    ]

## `CSS_files` Parameter

Parameter `CSS_files` (a list of directories and/or files) is used to explicitly define the CSS files to be used in the target HTML document.

When a directory is specified then _all_ files in the directory are currently copied into the target directory, regardless of their file extension.
In the new version, only files with extension `.css` will be copied.

The behavior remains unchanged if a file is specified. The file will be copied into the target directory, regardless of its extension.

This change has been discussed [here](https://github.com/pml-lang/pml-companion/issues/93).

## 'Text Block Syntax' Removed

The [Text Block Syntax](https://pdml-lang.dev/docs/extensions/reference_manual/index.html#raw_text_block_syntax) that was [deprecated](https://www.pml-lang.dev/docs/changelog/index.html#v_3_1_0_deprecated) in version 3.1.0. will no more be supported in raw text nodes.

The following nodes are affected by this change: `code`, `input`, `output`, `html`, `table_data`.

## `export_tags` Command

Command `export_tags` will be renamed to `export_meta_data`.

The default output file name will change from `pml_tags.json` to `pml_meta_data.json`.

The content of the JSON file will change as follows:

- Field `pml_tags` will be renamed to `pml_meta`.
- Field `tags` will be renamed to `nodes`.

## GitHub Documentation Repositories Merged Into a Single Repository``

The individual repositories for PML documentation (e.g. [User Manual](https://github.com/pml-lang/pml-userman)) will be removed, and their content (including issues and discussion) will be transferred to a single `pml-website` repository.

**Comment by pml-lang** [comment]
## Additional Changes

- Node `options` will have to be be defined _before_ node `doc` (not inside node `doc`). Hence this code:


```
[doc
    [options
        [TOC_title Aperçu]
    ]
    text
]
```
... will change to:
```
[options
    [TOC_title Aperçu]
]
[doc
    text
]
```

- Rename node `table_data` to `sim_table` (simple table). Note: In the future, node `sim_list` will be added to define simple lists with a much less verbose syntax.

- Integrate a 'PML Document Updater' into PMLC. This will enable users to automatically update existing PML documents that are not compatible with new breaking changes. This feature is not trivial to implement. PDML must be enhanced first. A first experimental implementation is scheduled to be published in version 4.1.0.

**Reply by slanden** [reply]
Just out of curiosity, was there a technical reason for moving `options` outside of `doc`?

This breaks PDML's ["Each document has exactly one root node."](https://pdml-lang.dev/docs/core/specification/index.html#node) rule.. Should the PDML specification be adjusted?

**Reply by tajmone** [reply]
Good point @slanden!. My impression is that since options concern command line options we're dealing with a sort of PMLC specific pre-processor layer, which is not truly part of the document tree (in fact these nodes don't seem to follow the usual order). But only @pml-lang can confirm this.

If that's the case, I might consider scoping `[option` nodes in Sublime PML as pre-processor directives, so they get a distinguished colour too, and better semantics.

**Reply by slanden** [reply]
That makes sense.

> "I might consider scoping [option nodes in Sublime PML as pre-processor directives" - @tajmone

I assigned similar scopes to all the PDML extension nodes in the VS Code extension, but I'm not sure how I feel about it honestly. The idea was that they have special meaning similar to programming languages, but on the other hand people may create their own nodes that also have meaning of some kind or *do* something special.. yet, their nodes will just get colored like regular nodes.

I'm just starting to wonder if perhaps consistency would be better. I'm undecided. I wish we could get PDML (and PML as well) in front of more eyes. Then, we could have meaningful polls for such questions.

**Reply by tajmone** [reply]
> I assigned similar scopes to all the PDML extension nodes in the VS Code extension, but I'm not sure how I feel about it honestly. The idea was that they have special meaning similar to programming languages, 

True, but then when it comes to markup syntaxes the scope naming conventions for TextMate-like editor (VSCode, Sublime Text, Atom, etc.) are rather obscure. The problem here is that these scopes were designed around popular languages (C/C++, Java, Python, etc.), and markup syntaxes like Markdown, Restructured Text, and that's about it — anything beyond those "classic" patterns and you're basically entering uncharted territory. 

Just think of BNF, EBNF, ABNF — there doesn't seem to be proper scoping names for these notations (and surely no guidelines), so much so that you won't find two BNF-related syntaxes using the same scope naming conventions, yet BNF grammars are the foundations for defining most languages, which IMO says it all about the "semantic soundness" of these scoping conventions.

> but on the other hand people may create their own nodes that also have meaning of some kind or do something special.. yet, their nodes will just get colored like regular nodes.

I'm not sure there's a way around this, unless you implement an LSP language server, in which case you're able to track custom node definitions project-wide, and then amend the generic scoping accordingly, dynamically.

My approach so far, for my PML and PDML packages for Sublime Text, has been to scope them as unknown nodes, which should cover both new nodes yet not covered by my packages, and user-defined nodes. This might be far from ideal, but at least it provides scope names that the user might leverage with custom plug-ins.

Thanks for sharing your thoughts on this. Since I also maintain editor packages for PDML and PML it's good if we share similar thought, which might help us improve our approaches. I'm pasting below the links to my packages, in case you want to peak at their scope-naming conventions:

* https://github.com/tajmone/Sublime-PDML
* https://github.com/tajmone/Sublime-PML

I haven't worked much on the PDML package (it's really bare bones right now) since I wanted to get a pulse of where PDML is going before taking big decisions. For the PML package, on the other hand, I've ventured into more detailed scoping, and I'm starting to gain insights into how the PML syntax should be approached (lenient parsing, alternative syntaxes, edge cases, etc.).

> I'm just starting to wonder if perhaps consistency would be better. I'm undecided. I wish we could get PDML (and PML as well) in front of more eyes. Then, we could have meaningful polls for such questions.

That's a hard decision. One could simply look at PDML/PML and consider every node as simply a tag, maybe distinguishing between inline and block nodes, and leave it at that. But PML is a markup syntax for editing, so it's a hybrid between HTML and markdown when it comes to scoping its nodes, and the way end users edit their documents with it.

I've opted to scope well established text-formatting nodes (italic, bold, etc.) along the lines of markdown, since most color themes should cover these markup scopes properly, rendering them in the proper style. For advanced elements like footnotes, there are no guidelines, so I'm just peeking at how third party packages implemented them in Markdown, AsciiDoc, and other markup syntaxes in other TextMate-like editors, to get a feeling of what conventions users might be already used to.

Rather than focusing on how the syntax will be colored in the editor, I'm investing more energy on coming up with consistent scope naming conventions and patters that will allow me and end users to target specific nodes or group of nodes with plugins code. E.g. I'm ensuring that `[ch` titles, `id` attributes and parameters are indexed, so that users can quickly navigate to their occurrences through the Goto Symbol menus, and I'm also applying some transformations to the indexed symbols, to improve their discovery and filtering (e.g. prefixing IDs with a `#`, like in HTML anchors, and so on).

I took me quite a while to come up with a smart way to meta-scope the whitespace between the opening tag and the beginning of its contents, the goal here being able to suggest completions based on context (i.e. only applying to the current node) and being able to distinguish when the cursor is positioned within the parenthesis grouping attributes, or in the whitespace just mentioned (where the node doesn't yet have any attributes), and ensuring that the whitespace is not thus scoped when the node has a group of attributes. Needless to say, this requires quite some RegEx juggling, and extra-contents to handle the various possibilities, but the final result is worth it since it provides end users with a better editing experience, where only relevant completions/snippets show up at any given place in the source document.

In the Discussions section of the Sublime PML repository I've posted some notes on how I approached these scoping challenges in the course of time, and how the changes in the syntax brought new challenges in this respect. Maybe these notes can be useful for your packages too. In any case, discussing and sharing similar dev notes on our repositories Discussions might be quite productive.

**Reply by tajmone** [reply]
## `[options` are Like YAML Headers in Markdown

Regarding the `[options` nodes group, I think we should look at them more like YAML metadata headers in Markdown documents — a different syntax and scope than the rest of the document. 

In the case of PML, the fact that `[options` nodes have a similar notation to the rest of PML nodes leads us to think of them as ordinary nodes, but at closer inspection we notice that they don't follow the usual pattern of PDML/PML nodes.

Since they refer to CLI commands, the scoping that comes to mind first is that of preprocessor, but if we look at them as a metadata header section then we can consider alternative scoping conventions (e.g. XML or HTML head sections).

**Reply by pml-lang** [reply]
> @slanden: was there a technical reason for moving options outside of doc?

There were three reasons:

1. Having the `options` node before the `doc` node makes it more useful/versatile.
   
    Example: The current options are all related to the target HTML code. But in the future we might add options for the PDML parser. This would be very challenging to achieve if the `options` node is parsed after parsing of the `doc` node has started already.

2. The `options` and `doc` tree represent semantically different data, therefore it's better to separate them.

    @tajmone put it like this:
    "Since options concern command line options we're dealing with a sort of PMLC specific pre-processor layer, which is not truly part of the document tree"
    And:
    "We should look at them more like YAML metadata headers in Markdown documents"

    Using PDML (and not a YAML frontmatter block, as is done in many frameworks that use Markdown, AsciiDoc, etc.) has two advantages:
    - The same syntax (PDML) is used for options and markup
    - The full power of PDML can be used to define options (insert from other file, scripting nodes, etc.)


3. Parsing and handling the `options` node inside the `doc` node required some ugly code to be written.
   
   It all ended up as a code smell. In such cases I lean back and ask myself: Are the data structures correct?
   Many times the answer is "No", and after fixing the data and code, the code usually smells like a rose.


> @slanden: This breaks PDML's "Each document has exactly one root node." rule. Should the PDML specification be adjusted?

Yes, the rule is that each document has exactly one root node.

No, the rule is not violated.

Generally speaking, consider the following PDML code:

```
[foo ...]
[bar ...]
[tar ...]
```

You can look at this code in two ways:

1. There are three PDML trees, with root nodes `foo`, `bar`, and `tar`.

2. There is a single tree with the root node omitted. The root node is implicitly defined (a virtual root node).

In the case of the `options` node I opted for the first approach, because that made more sense. PMLC first parses the `options` tree, then the `doc` tree. (Note: To achieve best performance, the parser uses a single character reader for the whole document. After parsing the `options` tree, the same character reader is used to continue reading characters that are part of the `doc` tree.)

The second option (virtual root node) can be useful for standalone config files (for example). Instead of writing:

```
[config
    [color green]
    [width 100]
]
```

... you can simply write:

```
[color green]
[width 100]
```

Internally, a root node is first inserted into the code, which can then be parsed as a normal PDML document with a single root.

Example: The full PDML implementation provides utility method [StringMapPdmlParser.parseFile](https://github.com/pdml-lang/full-pdml-impl/blob/7bcd6a50c96f1668468cc36182be13ea5023f8bc/pdml-utils/src/main/java/dev/pdml/utils/parser/StringMapPdmlParser.java#L56) with the following signature:

```
public static @Nullable Map<String, String> parseFile (
    @NotNull Path filePath,
    boolean hasRootNode ) throws IOException, TextErrorException {
```

You could use this method to parse a config file like the above one into a map of strings. If input argument `hasRootNode` is set to false then a virtual root node is implicitly injected.

> @tajmone: I might consider scoping [option nodes in Sublime PML as pre-processor directives, so they get a distinguished colour too, and better semantics.

Great!

> @slanden: I wish we could get PDML (and PML as well) in front of more eyes. Then, we could have meaningful polls for such questions.

Yes, that should really start happening this year. The word should be spread. PDML/PML is far from perfect, but probably "good enough" to start promoting it, so that more people can try it out, send feedback, and help to improve it. Ideas for promoting PDML/PML are very welcome.

**Comment by slanden** [comment]
> I'm investing more energy on coming up with consistent scope naming conventions and patters that will allow me and end users to target specific nodes or group of nodes with plugins code.

Your choice in being more specific in scoping might be a good one, especially if you're focusing on PML. I've intentionally kept things more generic for now. Thanks for sharing as well, it'll be interesting to come back after some time and see how the two approaches evolve.

To your point though, I can see value in that for sure. I was editing some HTML earlier and wish I could jump to different nodes.

> In the Discussions section of the Sublime PML repository I've posted some notes on how I approached these scoping challenges - @tajmone

Thanks for sharing. PDML doesn't require many scopes, but I plan to experiment with dialects/vocabularies like PML and reading up on your experiences would indeed be helpful. I'm more active on the PDML discussions, but I check these sometimes for more information or to see if someone else had thoughts I'm having.

Thanks @tajmone and @pml-lang, describing the `options` node that way helps me understand.

> Having the options node before the doc node makes it more useful/versatile.

I was thinking the same thing.

> You can look at this code in two ways:
> 1. There are three PDML trees, with root nodes foo, bar, and tar.
> 2. There is a single tree with the root node omitted. The root node is implicitly defined (a virtual root node).

I see the value in this, and your examples show what I consider natural uses of the language. I think this should be documented to make it more clear that this is possible. The current wording of the rule is black and white to me; clear, but evidently not the whole picture.

> Ideas for promoting PDML/PML are very welcome.

Hm.. maybe a youtube video of a workflow could be a good start; going over the capabilities, extensions, a few examples, then end with an example of a larger use of PDML, such as PML.

Or reverse that order; a series of PML videos, that sprinkle in little tidbits of PDML knowledge.

I've read at least two reddit threads about new markup languages, but be ready to see a larger margin of "ughhhh, another markup language" type commenters lol.

**Comment by pml-lang** [comment]
>  I think this should be documented to make it more clear that this is possible.

You're right. There are a few other features that are implemented already, but not yet documented.

> maybe a youtube video of a workflow ...

Thanks for the tip.

I've no experience with doing Youtube videos, therefore I was thinking more of an "Introduction to PDML" article (published on relevant sites) that focuses on advantages and shows examples of exclusive PDML features. The primary goal must be to explain _why_ people might want to have a closer look at PDML. Then do the same for PML, focusing on technical writers, because IMO they would benefit most from PML.

> I've read at least two reddit threads about new markup languages

Could you share the links, please?

**Reply by slanden** [reply]
> I was thinking more of an "Introduction to PDML" article ...

That could work just as well I think.

> Could you share the links, please?

Honestly, I don't remember exactly which ones. I was deep in a rabbit whole at the time, but I remember reading about the [KDL Document Language](https://kdl.dev/) (KDL; the acronym is recursive I think) and came across https://news.ycombinator.com/item?id=28510031 and also https://www.reddit.com/r/programming/comments/pmpy4h/the_kdl_document_language_an_alternative_to/.

**Reply by tajmone** [reply]
> I was thinking more of an "Introduction to PDML" article (published on relevant sites) that focuses on advantages and shows examples of exclusive PDML features. The primary goal must be to explain why people might want to have a closer look at PDML. Then do the same for PML, focusing on technical writers, because IMO they would benefit most from PML.

I don't really understand why this articles should be published elsewhere instead of on the PML official website.

For most software tools, a presentation of the tool, its features and why people should consider its adoption in their daily work, is usually part and parcel of the official website, alongside with documents explaining the project vision, roadmap, etc.

The official PML website _is_ the most relevant website for such material, and IMO constitute the foundation of any third party reviews of PML — and, I might add, the absence of such material might just as well be the primary cause for the lack of reviews at the present stage. For a potential reviewer, having some official material to start off with is much simpler, not only because it provides a solid foundation in terms of guidelines around which to structure and articulate his/her own review, but also because it would prevent the review from being perceived as opinionated.

**Reply by slanden** [reply]
The article and all educational materials *should* be published on the official website, and then they _can_ be re-posted on other sites. The purpose of publishing on other sites is to leverage their platforms for more exposure.

Take Medium.com for example, people are already there and people get newsletters in their email that *might* show it. I don't know what kind of ranking system they use, but the chance of someone just happening across the article is probably higher there than the official PML or PDML sites. If someone is already searching for "XML replacement" or something like that, they may find the PDML site thanks to one of Christian's older articles. That's my story at least.

Even still, the official websites are new and that also contributes to a lower rank in search engines.

**Reply by pml-lang** [reply]
> the KDL Document Language

Thanks for the links.

> The article and all educational materials should be published on the official website, and then they can be re-posted on other sites. The purpose of publishing on other sites is to leverage their platforms for more exposure.

Yes, exactly! Different people use different channels. Using multiple channels to promote the projects will increase the likelihood of finding new users.

> The official PML website is the most relevant website for such material

True. The websites should be improved, and there should be helpful articles targeted to new visitors who are curious about the benefits of these projects.

**Reply by tajmone** [reply]
> > The official PML website is the most relevant website for such material
> 
> True. The websites should be improved, and there should be helpful articles targeted to new visitors who are curious about the benefits of these projects.

What I also meant is that we're not dealing just with "articles" here, but documents which are fundamental for the project, and out of which articles will be based.

The PML website landing page should provide a clear presentation of what PML does and what people might expect to use for it. The website should also provide a _Vision_ document stating the long term goals of PML, what vision drives its design, and so on. A _Roadmap_ document should highlight the upcoming development steps, in terms of concrete changes, features, etc., so that potential users can get a clear picture of what to expect from PML in the upcoming years — which is what might motivate them to support the project, e.g. by funding it.

None of these documents can really be considered to be "articles", they are core documents of the project.

Currently, the PML website relies on various articles that were written to describe elsewhere what's happening with the project, but these can't compensate for the above mentioned documents.

Probably the priority right now is to reorganize the site contents in this respect, since so far the website has been "patched as we went along", the main focus being on developing the tools.
